ir/src/transition_constraints/graph.rs

```diff
@@ -0,0 +1,139 @@
+                let rhs = self.insert_expr(*rhs, trace_columns)?;
+                // add the expression
+                Ok(self.insert_op(Operation::Add(lhs, rhs)))
+            }
+            Expr::Subtract(lhs, rhs) => {
+                // add both subexpressions
+                let lhs = self.insert_expr(*lhs, trace_columns)?;
+                let rhs = self.insert_expr(*rhs, trace_columns)?;
+                // negate the right hand side
+                let rhs = self.insert_op(Operation::Neg(rhs));
+                // add the expression
+                Ok(self.insert_op(Operation::Add(lhs, rhs)))
+            }
+        }
+    }
+
+    /// Insert the operation with its degree and return its node index.
+    ///
+    /// If an identical node already exists, return that index instead.
+    fn insert_op(&mut self, operation: Operation) -> NodeIndex {
+        self.nodes
+            .iter()
+            .position(|n| *n.op() == operation)
+            .map_or_else(
+                || {
+                    // Create a new node.
+                    let index = self.nodes.len();
+                    // Get the degree of the node based on its child nodes
+                    let degree = match &operation {
+                        Operation::Constant(_)
+                        | Operation::CurrentRow(_)
+                        | Operation::NextRow(_) => 1_u8,
+                        Operation::Neg(index) | Operation::Inv(index) => self.node(index).degree(),
+                        Operation::Add(lhs, rhs) => {
+                            std::cmp::max(self.node(lhs).degree(), self.node(rhs).degree())
+                        }
+                        Operation::Mul(lhs, rhs) => {
+                            self.node(lhs).degree() + self.node(rhs).degree()
+                        }
+                        Operation::Exp(index, exp) => self.node(index).degree() * (*exp as u8),
+                    };
+
+                    self.nodes.push(Node {
+                        op: operation,
+                        degree,
+                    });
+
+                    NodeIndex(index)
+                },
+                |index| {
+                    // return existing node's index
+                    NodeIndex(index)
+                },
+            )
+    }
+}
+
+#[derive(Debug)]
+pub struct Node {
+    /// The operation represented by this node
+    op: Operation,
+    /// The degree of the expression represented by this node (and all children)
+    degree: u8,
+}
+
+impl Node {
+    pub fn op(&self) -> &Operation {
+        &self.op
+    }
+
+    pub fn degree(&self) -> u8 {
+        self.degree
+    }
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum Operation {
+    Constant(u64),
+    /// A column identifier for the current trace row containing the column index
+    CurrentRow(usize),
+    /// A column identifier for the next trace row containing the column index
+    NextRow(usize),
```

Comment:
Yeah, I'll update everything to reference Main. I had figured it would happen in the pass where we added Aux, but there's no good reason not to do it now.